src/Engine/Scene/SkinningComponent.cpp

```diff
@@ -0,0 +1,505 @@
+    m_meshIsPoly    = compMsg->canGet<PolyMesh>( getEntity(), m_meshName );
+
+    if ( hasSkel && hasRefPose && ( hasTriMesh || m_meshIsPoly ) )
+    {
+        m_renderObjectReader = compMsg->getterCallback<Index>( getEntity(), m_meshName );
+        m_skeletonGetter     = compMsg->getterCallback<Skeleton>( getEntity(), m_skelName );
+        if ( !m_meshIsPoly )
+        { m_triMeshWriter = compMsg->rwCallback<TriangleMesh>( getEntity(), m_meshName ); }
+        else
+        { m_polyMeshWriter = compMsg->rwCallback<PolyMesh>( getEntity(), m_meshName ); }
+
+        // copy mesh triangles and find duplicates for normal computation.
+        if ( !m_meshIsPoly ) { m_refData.m_referenceMesh = *m_triMeshWriter(); }
+        else
+        { m_refData.m_referenceMesh = triangulate( *m_polyMeshWriter() ); }
+        if ( !m_refData.m_referenceMesh.hasAttrib( tangentName ) &&
+             !m_refData.m_referenceMesh.hasAttrib( bitangentName ) )
+        {
+            const auto& normals = m_refData.m_referenceMesh.normals();
+            Vector3Array tangents( normals.size() );
+            Vector3Array bitangents( normals.size() );
+#pragma omp parallel for
+            for ( int i = 0; i < int( normals.size() ); ++i )
+            {
+                Core::Math::getOrthogonalVectors( normals[i], tangents[i], bitangents[i] );
+            }
+            m_refData.m_referenceMesh.addAttrib( tangentName, std::move( tangents ) );
+            m_refData.m_referenceMesh.addAttrib( bitangentName, std::move( bitangents ) );
+        }
+        else if ( !m_refData.m_referenceMesh.hasAttrib( tangentName ) )
+        {
+            const auto& normals = m_refData.m_referenceMesh.normals();
+            const auto& bH = m_refData.m_referenceMesh.getAttribHandle<Vector3>( bitangentName );
+            const auto& bitangents = m_refData.m_referenceMesh.getAttrib( bH ).data();
+            Vector3Array tangents( normals.size() );
+#pragma omp parallel for
+            for ( int i = 0; i < int( normals.size() ); ++i )
+            {
+                tangents[i] = bitangents[i].cross( normals[i] );
+            }
+            m_refData.m_referenceMesh.addAttrib( tangentName, std::move( tangents ) );
+        }
+        else if ( !m_refData.m_referenceMesh.hasAttrib( bitangentName ) )
+        {
+            const auto& normals = m_refData.m_referenceMesh.normals();
+            const auto& tH      = m_refData.m_referenceMesh.getAttribHandle<Vector3>( tangentName );
+            const auto& tangents = m_refData.m_referenceMesh.getAttrib( tH ).data();
+            Vector3Array bitangents( normals.size() );
+#pragma omp parallel for
+            for ( int i = 0; i < int( normals.size() ); ++i )
+            {
+                bitangents[i] = normals[i].cross( tangents[i] );
+            }
+            m_refData.m_referenceMesh.addAttrib( bitangentName, std::move( bitangents ) );
+        }
```

Comment:
Shall we have tangent and bitangent vectors computation in `Core` ?
This might be useful for other purposes than just skinning.
Also, there might be better ways to compute them, like consistent and continuous ways using curvature or something else?

Reply:
As this operation is done also on some fileloaders, yes, it is planned to bring such algorithms to the core. This will be done in a next PR as soon as we found a good design for this.